File: 数据结构/动态规划练习/DPtest_3/DPtest_3/ViewController.m
//
//  ViewController.m
//  DPtest_3
//
//  Created by 俊祥 on 2018/11/17.
//  Copyright © 2018年 俊祥. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()

@end

@implementation ViewController
// 动态规划 题目 求两个字符串的最大公共子串
- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    NSString *maxString = [self maxCommonString:@"hghahasjfh" andString:@"agasjfjkhhga"];
    NSLog(@"%@",maxString);
}

- (NSString *)maxCommonString:(NSString *)s1 andString:(NSString *)s2 {
    NSString *res = @"";
    
    if (s1.length <= 0 || s2.length <= 0) {
        return res;
    }
    
    int max = 0,m = (int)s1.length,n = (int)s2.length;
    
    int dp[100][100] = {0}; // 记录最大公共子串的长度
    for (NSInteger i = 0; i < m; i++) {
        for (NSInteger j = 0; j < n; j++) {
            if ([s1 characterAtIndex:i] == [s2 characterAtIndex:j]) {
                // 说明相等了
                if (i == 0 || j == 0) {
                    dp[i][j] = 0;
                } else {
                    dp[i][j] = dp[i-1][j-1] + 1;
                }
                // 记录最大长度和子串
                if (dp[i][j] > max) {
                    max = dp[i][j];
                    NSLog(@"%@",@(max));
                    res = [s1 substringWithRange:(NSRange){i-max+1,max}];
                }
            }
        }
    }
    
    return res;
}

@end

File: algorithms/algorithms/ViewController.swift
//
//  ViewController.swift
//  algorithms
//
//  Created by 俊祥 on 2017/6/24.
//  Copyright © 2017年 俊祥. All rights reserved.
//

import UIKit

class ViewController: UIViewController,UITableViewDelegate,UITableViewDataSource {

    let cellIdentifer:String = "cellIdentifer"
    var datas = [12,23,42,21,52,53,3,51,65,34,64]
    
    var sorts = ["Bubble"]
     
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        self.title = "算法练习"
        view.backgroundColor = .white
        findView()
    }
    
    func findView() {
        let table = UITableView(frame: JXScreenRect)
        table.delegate = self
        table.dataSource = self
        view .addSubview(table)
    }
    
    func gotoNextC() {
        
        
        let bubVC = BubbleSortController()
        bubVC.datas = datas
        self.navigationController?.pushViewController(bubVC, animated: true)
        
    }
    
    /// tableViewDelegate. tableViewDataSource
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return sorts.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifer) ?? UITableViewCell(style: UITableViewCellStyle.default, reuseIdentifier: cellIdentifer)
        
        cell.textLabel?.text = sorts[indexPath.row]
        
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        gotoNextC()
    }
    
    
}

